src/crypto.js

```diff
@@ -0,0 +1,266 @@
+ @property {sring} salt - The salt used to derive the key (format: hex string)
+ @property {Number} iterations - The iteration # used during key derivation
+ */
+
+const requiredParameterHashedPassphrase = ['salt', 'iterations', 'storedHash', 'hashAlgo']
+
+const _checkPassphrase = (passphrase) => {
+  if (typeof passphrase !== 'string' || passphrase === '') {
+    throw generateError(ERRORS.NOPASSPHRASE)
+  }
+}
+
+const _checkCryptokey = (key) => {
+  if (!key.type || key.type !== 'secret') {
+    throw generateError(ERRORS.NOCRYPTOKEY)
+  }
+}
+
+/**
+ * Generate a PBKDF2 derived key based on user given passPhrase
+ *
+ * @param {string | arrayBuffer} passPhrase The passphrase that is used to derive the key
+ * @param {arrayBuffer} [salt] The salt
+ * @param {Number} [iterations] The iterations number
+ * @param {string} [hash] The hash function used for derivation
+ * @returns {Promise<Uint8Array>}   A promise that contains the derived key
+ */
+const deriveBits = (passPhrase, salt, iterations, hash) => {
+  // Always specify a strong salt
+  if (iterations < 10000) { console.warn('The iteration number is less than 10000, increase it !') }
+
+  return window.crypto.subtle.importKey(
+    'raw',
+    (typeof passPhrase === 'string') ? Buffer.from(passPhrase) : passPhrase,
+    'PBKDF2',
+    false,
+    ['deriveBits', 'deriveKey']
+  )
+    .then(baseKey => {
+      return window.crypto.subtle.deriveBits({
+        name: 'PBKDF2',
+        salt: salt || new Uint8Array([]),
+        iterations: iterations || 100000,
+        hash: hash || 'SHA-256'
+      }, baseKey, 128)
+    })
+    .then(derivedKey => new Uint8Array(derivedKey))
+}
+
+/**
+ * Hash of a string or arrayBuffer
+ *
+ * @param {string | arrayBuffer} msg The message
+ * @param {string} [type] The hash name (SHA-256 by default)
+ * @returns {Promise<Uint8Array>}   A promise that contains the hash as a Uint8Array
+ */
+const hash256 = (msg, type = 'SHA-256') => {
+  return window.crypto.subtle.digest(
+    {
+      name: 'SHA-256'
+    },
+    (typeof msg === 'string') ? Buffer.from(msg) : msg
+  )
+    .then(digest => new Uint8Array(digest))
+}
+
+/**
+ * Derive a passphrase and return the object to store
+ *
+ * @param {string | arrayBuffer} passPhrase The passphrase that is used to derive the key
+ * @returns {Promise<HashedPassphrase>}   A promise that contains the derived key
+ */
+const derivePassphrase = (passPhrase) => {
+  _checkPassphrase(passPhrase)
+  let hashedPassphrase = {}
+  const salt = genRandomBuffer(16)
+  const iterations = 100000
+  hashedPassphrase.salt = Buffer.from(salt).toString('hex')
+  hashedPassphrase.iterations = iterations
+  hashedPassphrase.hashAlgo = 'SHA-256'
+  return deriveBitsAndHash(passPhrase, salt, iterations)
+    .then(hashedValue => {
+      hashedPassphrase.storedHash = Buffer.from(hashedValue).toString('hex')
+      return hashedPassphrase
+    })
+    .catch(err => console.log(err)
```

Comment:
Please use console.error in a catch block

Reply:
ok